Adding functionality to override the pandas default na values.

SFuller4 opened the pull request:
This is directly solving this issue [https://github.com/dataiku/dss-plugin-neo4j/issues/29](Issue #29).

StanislasGuinel commented on custom-recipes/neo4j-export-nodes/recipe.json:
```diff
@@ -110,6 +110,28 @@
+            "name": "keep_default_na",
+            "label": "Keep Default null Values",
+            "type": "BOOLEAN",
+            "description": "By default the following values are interpreted as NaN: ‘’, ‘#N/A’, ‘#N/A N/A’, ‘#NA’, ‘-1.#IND’, ‘-1.#QNAN’, ‘-NaN’, ‘-nan’, ‘1.#IND’, ‘1.#QNAN’, ‘<NA>’, ‘N/A’, ‘NA’, ‘NULL’, ‘NaN’, ‘None’, ‘n/a’, ‘nan’, ‘null’",
```
Instead of listing all the default values, just use a link to pandas documentation so that the description is smaller and won't be outdated.

StanislasGuinel commented on custom-recipes/neo4j-export-nodes/recipe.json:
```diff
@@ -110,6 +110,28 @@
+        {
+            "label": "Handling Null Values",
+            "type": "SEPARATOR",
+            "description": "Whether or not to include the default NaN values when parsing the data. Depending on whether na_values is passed in, the behavior is as follows: If keep_default_na is True, and na_values are specified, na_values is appended to the default NaN values used for parsing. If keep_default_na is True, and na_values are not specified, only the default NaN values are used for p ...
```
Prefer using a smaller description that can fit on at most 2 lines. The complete description can be added later in the plugin documentation page.

StanislasGuinel commented on custom-recipes/neo4j-export-nodes/recipe.json:
```diff
@@ -110,6 +110,28 @@
             "defaultValue": false
         },
+        {
+            "label": "Handling Null Values",
```
This new section should go at the bottom (below the `Export parameters function`)

StanislasGuinel reviewed (commented):
This recipes are broken: they fail with the following error message: `TypeError: __init__() got an unexpected keyword argument 'na_values'`. You need to add the new params to the `NodesExportParams` and `RelationshipsExportParams` classes.

You must also edit the `CHANGELOG.md` file.

Finally, you should prepare a description (in a comment of this PR for instance) that we will add to the plugin documentation page when we publish your changes.

SFuller4 commented:
## Handling NULL Values

* **Custom NULL Values**: Strings potentially found in the dataset, intended to be read as NULL values before being sent to Neo4j. If there are no strings supplied, empty cells will be replaced with the string 'null'. Only a list of strings that will apply to the entire dataset is supported through this plugin. 

* **Add default strings**: If selected, the default strings that pandas uses will be added to the list of custom NULL values. You can find information about what those default strings are here [Pandas.to_table()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_table.html) in the na_values section. 

### Usage Notes
* If Add default strings is True, and Custom NULL Values are specified, Custom NULL Values is appended to the default NULL values used for parsing.
* If Add default strings is True, and Custom NULL Values are not specified, only the default NULL values are used for parsing.
* If Add default strings is False, and Custom NULL Values are specified, only the NULL values specified Custom NULL Values are used for parsing.
* If Add default strings is False, and Custom NULL Values are not specified, no strings will be parsed as NULL.

SFuller4 commented:
@StanislasGuinel I updated the request and resolved your comments. I've also added a comment that has the data for the documentation page.
![image](https://user-images.githubusercontent.com/39442631/235976895-dc77777b-86d0-4975-9ccf-5d6b2d093a1f.png)


I also added the new parameters to the pytest file, although I was unable to perform the pytest locally. 

Let me know if there are any other problems so I can help get this in place as soon as possible!

StanislasGuinel commented:
Thanks for the comment for the documentation page !

Did you test that your feature work ? It still fails (`TypeError: create_dataframe_iterator() got an unexpected keyword argument 'na_value'`)

StanislasGuinel commented:
Also, there are many line diffs in this PR that shouldn't be there, I guess you used a formatter, but that makes the code review much harder. Could you revert all the useless diffs so that the review can only focus on the new feature code ?

StanislasGuinel commented on python-lib/dku_neo4j/neo4j_handle.py:
```diff
@@ -518,6 +524,9 @@ def __init__(
         if self.relationship_id_column:
             self.used_columns.append(self.relationship_id_column)
 
+        self.na_values = na_values
+        self.keep_default_na = keep_default_na
```
to be consistent with the other expert parameters, make sure to only use these 2 params when expert mode is activated
```suggestion
        self.na_values = na_values if expert_mode else None
        self.keep_default_na = keep_default_na if expert_mode else True
```

StanislasGuinel commented on custom-recipes/neo4j-export-nodes/recipe.json:
```diff
@@ -148,6 +148,27 @@
+        },
+        {
+            "name": "keep_default_na",
+            "label": "Add default strings",
```
```suggestion
            "label": "Include default NaN values",
```

StanislasGuinel commented on custom-recipes/neo4j-export-nodes/recipe.json:
```diff
@@ -148,6 +148,27 @@
             "visibilityCondition": "model.expert_mode"
+        },
+        {
+            "label": "Handling NULL Values",
```
Let's use the wording from pandas
```suggestion
            "label": "Handling NaN Values",
```

StanislasGuinel commented on custom-recipes/neo4j-export-nodes/recipe.json:
```diff
@@ -148,6 +148,27 @@
+            "name": "keep_default_na",
+            "label": "Add default strings",
+            "type": "BOOLEAN",
+            "description": "In addition to specified custom values, use the standard pre-defined strings found at https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_table.html",
```
```suggestion
            "description": "Whether or not to include the default pandas NaN values when parsing the data",
```

StanislasGuinel commented on custom-recipes/neo4j-export-nodes/recipe.json:
```diff
@@ -148,6 +148,27 @@
+            "name": "na_values",
+            "label": "Custom NULL Values",
+            "type": "STRINGS",
+            "description": "Specify custom string values to be interpreted as NULL prior to the export to Neo4j. No values supplied results in no NULL values, thus empty cells will be replaced with the string 'null'",
```
Let's keep the description in the UI really simple and just copy what pandas did
```suggestion
            "description": "Additional strings to recognize as NaN values prior to the export to Neo4j",
```

StanislasGuinel commented on custom-recipes/neo4j-export-nodes/recipe.json:
```diff
@@ -148,6 +148,27 @@
+        },
+        {
+            "name": "na_values",
+            "label": "Custom NULL Values",
```
```suggestion
            "label": "Additional NaN values",
```

StanislasGuinel commented on custom-recipes/neo4j-export-nodes/recipe.json:
```diff
@@ -148,6 +148,27 @@
+        },
+        {
+            "label": "Handling NULL Values",
+            "type": "SEPARATOR",
+            "visibilityCondition": "model.expert_mode"
```
Actually we should add the link to pandas in a description field of the separator because links in a separator are recognised as link (and not just string).
```suggestion
            "type": "SEPARATOR",
            "description": "More detailed documentation: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_table.html",
            "visibilityCondition": "model.expert_mode"
```

StanislasGuinel reviewed (commented):
I tested it, it works well !

Note: to run the unit-tests, you just need to run `make unit-tests` at the root of the plugin directory.

StanislasGuinel commented on CHANGELOG.md:
```diff
@@ -1,5 +1,9 @@
 
+## [Version 1.4.1](https://github.com/dataiku/dss-plugin-neo4j/tree/v1.4.1) - Bugfix release - 2023-05
+
+- Added Expert Params Custom Null Value and Add default Strings to override the default behavior of pandas.read_table and read_csv
```
```suggestion
- Specify whether or not to include the default NaN values and use additional strings to recognize as NaN values when exporting data to Neo4j
```

SFuller4 commented:
Those changes are good with me! I went ahead and updated my previous comment for the documentation to use NaN instead of NULL as well, to stay consistent.

SFuller4 commented:
Brought the Export Relationships json in line with Export Nodes json.

StanislasGuinel reviewed (approved):
All good, I'll merge this PR, and then run our CI.

SFuller4 commented:
Thanks for the help @StanislasGuinel ! Everything updated and looks great from our side!
